python/x402_a2a/core/wallet.py

```diff
@@ -27,85 +34,225 @@
+    # --- Get EIP-712 domain info from the contract ---
+    chain_id = w3.eth.chain_id
+    token_name = asset_contract.functions.name().call()
+    token_version = asset_contract.functions.version().call()
+
+    # Set up logging
+    logging.basicConfig(level=logging.INFO)
+
+    # --- 2. Create the exact data structure for signing ---
+    typed_data = get_transfer_with_auth_typed_data(
+        from_=auth_data["from_"],
+        to=auth_data["to"],
+        value=auth_data["value"],
+        valid_after=auth_data["valid_after"],
+        valid_before=auth_data["valid_before"],
+        nonce=auth_data["nonce"],
+        chain_id=chain_id,
+        contract_address=ASSET_CONTRACT_ADDRESS,
+        token_name=token_name,
+        token_version=token_version,
+    )
 
-    signed_base64 = sign_payment_header(
-        account=account, payment_requirements=requirements, header=unsigned_payload
+    logging.info("--- SIGNING DEBUG DATA (Client) ---")
+    logging.info(f"domain: {json.dumps(typed_data['domain'], indent=2)}")
+    logging.info(
+        f"message: {json.dumps({k: (f'0x{v.hex()}' if isinstance(v, bytes) else v) for k, v in typed_data['message'].items()}, indent=2)}"
     )
+    logging.info("-----------------------------------")
 
-    # Step 3: Decode back to proper PaymentPayload structure
-    signed_payload = decode_payment(signed_base64)
+    # --- 3. Sign THIS EXACT DATA OBJECT ---
+    signable_message = encode_typed_data(full_message=typed_data)
+    signed_message = account.sign_message(signable_message)
 
-    # Step 4: Convert to our PaymentPayload types
-    auth_data = signed_payload["payload"]["authorization"]
+    # --- 4. Construct the final payload using the SAME authorization data ---
     authorization = EIP3009Authorization(
-        from_=auth_data["from"],
+        from_=auth_data["from_"],
         to=auth_data["to"],
-        value=auth_data["value"],
-        valid_after=auth_data["validAfter"],
-        valid_before=auth_data["validBefore"],
-        nonce=auth_data["nonce"],
+        value=str(auth_data["value"]),
+        valid_after=str(auth_data["valid_after"]),
+        valid_before=str(auth_data["valid_before"]),
+        nonce=f"0x{auth_data['nonce'].hex()}",
     )
 
+    # The signature is a single bytes object, but some systems expect it as a hex string
+    # with r, s, and v components concatenated.
+    signature_hex = f"0x{signed_message.r.to_bytes(32, 'big').hex()}{signed_message.s.to_bytes(32, 'big').hex()}{signed_message.v:02x}"
+
     exact_payload = ExactPaymentPayload(
-        signature=signed_payload["payload"]["signature"], authorization=authorization
+        signature=signature_hex, authorization=authorization
     )
 
     return PaymentPayload(
-        x402_version=signed_payload["x402Version"],
-        scheme=signed_payload["scheme"],
-        network=signed_payload["network"],
+        x402_version=x402_VERSION,
+        scheme="exact",
+        network=requirements.network,
```

Comment:
Are we intentionally only allowing exact scheme?

This is in the core library, so it'll be a breaking change, if I understand correctly.